internal/provider/resource_apply_module.go

```diff
@@ -554,6 +683,125 @@ func (t *applyModuleResource) getCurrentApplied(ctx context.Context,
 	return nil
 }
 
+// extractRunError extracts the error from a run's logs (if the run errored out).
+func (t *applyModuleResource) extractRunError(ctx context.Context, run *sdktypes.Run) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if run.Status == sdktypes.RunErrored {
+		var jobID string
+
+		switch {
+		case run.Apply != nil:
+			if run.Apply.CurrentJobID != nil {
+				jobID = *run.Apply.CurrentJobID
+			}
+		case run.Plan != nil:
+			if run.Plan.CurrentJobID != nil {
+				jobID = *run.Plan.CurrentJobID
+			}
+		}
+
+		if jobID == "" {
+			diags.AddWarning("Run status is errored, but no job ID found", "")
+			return diags
+		}
+
+		// Must get the job to know the size of the logs to paginate in reverse.
+		job, err := t.client.Job.GetJob(ctx, &sdktypes.GetJobInput{
+			ID: jobID,
+		})
+		if err != nil {
+			diags.AddError("Failed to get job", err.Error())
+			return diags
+		}
+
+		// Get the logs from the end.  There will most likely be a smaller chunk at the beginning.
+		allLogs := ""
+		currentStart := int32(job.LogSize) - getLogsChunkSize
+		nextChunkSize := int32(getLogsChunkSize)
+		if currentStart < 0 {
+			// Only one chunk to read.
+			currentStart = 0
+			nextChunkSize = int32(job.LogSize)
+		}
+		for {
+			logs, err := t.client.Job.GetJobLogs(ctx, &sdktypes.GetJobLogsInput{
+				JobID: jobID,
+				Start: currentStart,
+				Limit: &nextChunkSize,
+			})
+			if err != nil {
+				diags.AddError("Failed to get job logs", err.Error())
+				return diags
+			}
+
+			// Workaround: The API returns one more character than asked for.
+			newLogs := logs.Logs
+			if len(newLogs) > int(nextChunkSize) {
+				newLogs = newLogs[:nextChunkSize]
+			}
+
+			allLogs = newLogs + allLogs
+			if strings.HasPrefix(allLogs, "Error: ") {
+				// Found the error, so break out of the loop.
+				break
+			}
+
+			if currentStart == 0 {
+				// No error found, and we've read the whole log.
+				break
+			}
+
+			if currentStart >= getLogsChunkSize {
+				currentStart -= getLogsChunkSize
+			} else {
+				// A smaller chunk at the beginning.
+				nextChunkSize = currentStart
+				currentStart = 0
+			}
+		}
+
+		// Find the first mention of "error" in the logs.
+		splitLogs := strings.Split(allLogs, "\n")
+		foundIx := -1
```

Comment:
This shouldn't need to split the logs since you can find the index of `Error:` and `Created new state version` and then splice out the error using the two indexes.